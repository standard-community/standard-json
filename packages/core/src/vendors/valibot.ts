import type { JSONSchema7 } from "json-schema";
import { toJsonSchema as vToJsonSchema } from "@valibot/to-json-schema";
import type { BaseSchema } from "valibot";

export const toJsonSchema = (
  schema: BaseSchema<any, any, any>,
  options?: Record<string, unknown>,
) => vToJsonSchema(schema, options) as JSONSchema7;
