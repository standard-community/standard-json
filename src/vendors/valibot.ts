import { errorMessageWrapper, type ToJsonSchemaFn } from "./utils.js";

export default async function getToJsonSchemaFn(): Promise<ToJsonSchemaFn> {
  const { toJsonSchema } = await import("@valibot/to-json-schema").catch(() => {
    throw new Error(
      errorMessageWrapper('Missing dependencies "@valibot/to-json-schema".'),
    );
  });
  return toJsonSchema as ToJsonSchemaFn;
}
