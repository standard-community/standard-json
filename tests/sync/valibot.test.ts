import { toJsonSchema as valibotToJsonSchema } from "@valibot/to-json-schema";
import * as v from "valibot";
import { describe, expect, it } from "vitest";

import { loadVendor, toJsonSchema } from "~/index.js";
import type { ToJsonSchemaFn } from "~/vendors/utils.js";

describe("sync check", () => {
  it("valibot", () => {
    // Loading the vendor
    loadVendor("valibot", valibotToJsonSchema as ToJsonSchemaFn);

    const schema = v.pipe(
      v.object({
        myString: v.string(),
        myUnion: v.union([v.number(), v.boolean()]),
      }),
      v.description("My neat object schema"),
    );

    const jsonSchema = toJsonSchema.sync(schema);
    expect(jsonSchema).toMatchSnapshot();
  });
});
