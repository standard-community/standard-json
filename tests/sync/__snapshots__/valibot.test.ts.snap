// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sync check > valibot 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;
